pkg/services/ngalert/migration/service.go

```diff
@@ -49,9 +69,464 @@ func ProvideService(
+							pair.ContactPointUpgrade = fromContactPointUpgrade(recv, route)
+						}
+					}
+				}
+				resState.MigratedChannels = append(resState.MigratedChannels, pair)
+			}
+		}
+
+		// We preload information in bulk for performance reasons.
+		mappedRules, err := ms.migrationStore.GetSlimAlertRules(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("get all alert rules: %w", err)
+		}
+
+		mappedDashboardAlerts, err := ms.migrationStore.GetSlimOrgDashboardAlerts(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("get dashboard alerts: %w", err)
+		}
+
+		dashIDInfo, err := ms.migrationStore.GetSlimDashboards(ctx, orgID)
+		if err != nil {
+			return fmt.Errorf("get dashboards: %w", err)
+		}
+		dashUIDInfo := make(map[string]migrationStore.SlimDashboard, len(dashIDInfo))
+		for _, info := range dashIDInfo {
+			dashUIDInfo[info.UID] = info
+		}
+
+		existingDashboards := make(map[int64]struct{})
+		for dashboardID, alerts := range mappedDashboardAlerts {
+			existingDashboards[dashboardID] = struct{}{}
+			mDu := &definitions.DashboardUpgrade{
+				MigratedAlerts: make([]*definitions.AlertPair, 0),
+				DashboardID:    dashboardID,
+			}
+
+			dashInfo, ok := dashIDInfo[dashboardID]
+			if ok {
+				folderInfo := dashIDInfo[dashInfo.FolderID]
+				mDu.DashboardUID = dashInfo.UID
+				mDu.DashboardName = dashInfo.Title
+				mDu.FolderUID = folderInfo.UID
+				mDu.FolderName = folderInfo.Title
+				mDu.Provisioned = dashInfo.Provisioned
+			}
+
+			du, ok := dState.MigratedDashboards[dashboardID]
+			if !ok {
+				mDu.Error = "dashboard not upgraded"
+				// Empty dashboard upgrade, to simplify logic below.
+				du = &migrationStore.DashboardUpgrade{
+					DashboardID:    dashboardID,
+					MigratedAlerts: make(map[int64]*migrationStore.AlertPair),
+				}
+			}
+			mDu.Warning = du.Warning
+
+			if du.AlertFolderUID != "" {
+				newFolderInfo := dashUIDInfo[du.AlertFolderUID]
+				mDu.NewFolderUID = du.AlertFolderUID
+				mDu.NewFolderName = newFolderInfo.Title
+			}
+
+			existingAlerts := make(map[int64]struct{})
+			for _, a := range alerts {
+				existingAlerts[a.ID] = struct{}{}
+				pair := &definitions.AlertPair{
+					LegacyAlert: fromSlimAlert(a),
+					Error:       "alert not upgraded",
+				}
+
+				if p, ok := du.MigratedAlerts[a.PanelID]; ok {
+					pair.Error = p.Error
+					if p.NewRuleUID != "" {
+						if rule, ok := mappedRules[p.NewRuleUID]; ok {
+							var sendTo = make([]string, 0)
+							for _, m := range amConfig.Match(extractLabels(rule, mDu.NewFolderName)) {
+								sendTo = append(sendTo, m.RouteOpts.Receiver)
+							}
+							pair.AlertRule = fromSlimAlertRule(rule, sendTo)
+						} else {
+							// We could potentially set an error here, but it's not really an error. It just means that the
+							// user deleted the migrated rule after the migration. This could just as easily be intentional.
+							ms.log.Info("Could not find rule for migrated alert", "alertID", a.ID, "ruleUID", p.NewRuleUID)
```

Comment:
I've seen in some places that we use `Id` and `Uid` instead. I think it would be great to be consistent in a new code.